go/vt/schemadiff/schema_diff.go

```diff
@@ -0,0 +1,247 @@
+/*
+Copyright 2023 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package schemadiff
+
+import (
+	"fmt"
+
+	"vitess.io/vitess/go/mathutil"
+)
+
+type DiffDepType int
+
+// diff dependencies in increasing restriction severity
+const (
+	DiffDepNone DiffDepType = iota // not a dependency
+	DiffDepOrderUnknown
+	DiffDepInOrderCompletion
+	DiffDepSequentialExecution
+)
+
+// DiffDep indicates a dependency between two diffs, and the type of that dependency
+type DiffDep struct {
+	diff    EntityDiff
+	depDiff EntityDiff // depends on the above diff
+	depType DiffDepType
+}
+
+func NewDiffDep(diff EntityDiff, depDiff EntityDiff, depType DiffDepType) *DiffDep {
+	return &DiffDep{
+		diff:    diff,
+		depDiff: depDiff,
+		depType: depType,
+	}
+}
```

Comment:
This isn't concurrency-safe, right? That seems fine but IMO we should comment on that.

Reply:
Correct! Clarified in code comment on `SchemaDiff` struct.